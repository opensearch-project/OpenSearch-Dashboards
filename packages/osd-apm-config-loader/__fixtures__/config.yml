pid:
  enabled: true
  file: '/var/run/opensearch_dashboards.pid'
  obj: { val: 3 }
  arr: [1]
  empty_obj: {}
  empty_arr: []
obj: { val: 3 }
arr: [1, 2]
empty_obj: {}
empty_arr: []
