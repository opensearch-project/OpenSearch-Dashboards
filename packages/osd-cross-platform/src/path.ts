/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { execSync } from 'child_process';
import { basename, normalize, resolve } from 'path';
import {
  realpathSync as nativeRealpathSync,
  openSync,
  closeSync,
  existsSync,
  unlinkSync,
} from 'fs';

export const NAMESPACE_PREFIX = process.platform === 'win32' ? '\\\\?\\' : '';

/**
 * Get a standardized reference to a path
 * @param {string} path - the path to standardize
 * @param {boolean} [usePosix=true] - produce a posix reference
 * @param {boolean} [escapedBackslashes=true] - on Windows, double-backslash the reference
 * @param {boolean} [returnUNC=false] - produce an extended reference
 */
export const standardize = (
  path: string,
  usePosix: boolean = true,
  escapedBackslashes: boolean = true,
  returnUNC: boolean = false
) => {
  // Force os-dependant separators
  const normal = normalize(path);

  // Filter out in-browser executions as well as non-windows ones
  if (process?.platform !== 'win32') return normal;

  if (usePosix) return normal.replace(/\\/g, '/');
  else if (escapedBackslashes) return normal.replace(/\\/g, '\\\\');
  else if (returnUNC) return '\\\\?\\' + normal;
  return normal;
};

/**
 * Windows-only function that uses PowerShell to calculate the full path
 * @param {string} path
 * @private
 */
const getFullPathSync = (path: string) => {
  if (process.platform !== 'win32') return path;

  try {
    const fullName = execSync(`powershell "(Get-Item -LiteralPath '${path}').FullName"`, {
      encoding: 'utf8',
    })?.trim?.();

    // Make sure we got something back
    if (fullName?.length > 2) return fullName;
  } catch (ex) {
    // Do nothing
  }

  return path;
};

/**
 * Windows-only function that uses PowerShell and Com Object to calculate the 8.3 path
 * @param {string} path
 * @private
 */
const getShortPathSync = (path: string) => {
  if (process.platform !== 'win32') return path;

  try {
    const shortPath = execSync(
      `powershell "$FSO = New-Object -ComObject Scripting.FileSystemObject; $O = (Get-Item -LiteralPath '${path}'); if ($O.PSIsContainer) { $FSO.GetFolder($O.FullName).ShortPath } else { $FSO.GetFile($O.FullName).ShortPath }"`,
      {
        encoding: 'utf8',
      }
    )?.trim?.();

    // Make sure we got something back
    if (shortPath?.length > 2) return shortPath;
  } catch (ex) {
    // Do nothing
  }

  return path;
};

/**
 * Checks if Windows 8.3 short names are supported on the volume of the given path
 * @param {string} [path='.'] - the path to examine
 */
export const shortNamesSupportedSync = (path: string = '.') => {
  if (process.platform !== 'win32') return false;

  const testFileName = '.___osd-cross-platform-test.file';
  const file = resolve(path, testFileName);

  // Create a test file if it doesn't exist
  if (!existsSync(file)) closeSync(openSync(file, 'w'));

  // If the returned value's basename is not the same as the requested file name, it must be a short name
  const foundShortName = basename(getShortPathSync(file)) !== testFileName;

  // Cleanup
  unlinkSync(file);

  return foundShortName;
};

/**
 * @borrows shortNamesSupportedSync
 */
export const shortNameSupportedSync = shortNamesSupportedSync;

/**
 * Get the full pathname
 * @param {string} path - the path to resolve
 */
export const resolveToFullPathSync = (path: string) => getFullPathSync(resolve(path));

/**
 * @borrows resolveToFullPathSync
 */
export const resolveToFullNameSync = resolveToFullPathSync;

/**
 * Get the short pathname
 * @param {string} path - the path to resolve
 */
export const resolveToShortPathSync = (path: string) => getShortPathSync(resolve(path));

/**
 * @borrows resolveToShortPathSync
 */
export const resolveToShortNameSync = resolveToShortPathSync;

/**
 * Get the canonical pathname
 * @param {string} path - the path to resolve
 */
export const realPathSync = (path: string) => getFullPathSync(nativeRealpathSync(path, 'utf8'));

/**
 * @borrows realPathSync
 */
export const realpathSync = realPathSync;

/**
 * Get the canonical pathname
 * @param {string} path - the path to resolve
 */
export const realShortPathSync = (path: string) =>
  getShortPathSync(nativeRealpathSync(path, 'utf8'));

/**
 * @borrows realShortPathSync
 */
export const realshortpathSync = realShortPathSync;
