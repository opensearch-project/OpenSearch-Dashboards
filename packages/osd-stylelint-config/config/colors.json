{
  "background-color": {
    "buttonTODOTHISANEXAMPLE": [
      {
        "approved": "$euiColorPrimary",
        "rejected": [
          "blueTODOTHISANEXAMPLE"
        ]
      },
      {
        "approved": "$euiColorWarning",
        "rejected": [
          "redTODOTHISANEXAMPLE"
        ]
      }
    ]
  }
}
