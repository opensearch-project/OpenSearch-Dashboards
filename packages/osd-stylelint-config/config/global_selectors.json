{
  "#opensearch-dashboards-body": {
    "approved": [
      "src/core/public/rendering/_base.scss"
    ]
  },
  ".app-wrapper": {
    "approved": [
      "src/core/public/rendering/_base.scss"
    ]
  },
  "/\\.[eo]ui/": {
    "approved": [
      "examples/expressions_example/public/index.scss",
      "src/core/public/styles/_base.scss",
      "src/plugins/vis_default_editor/public/_sidebar.scss",
      "src/core/public/core_app/styles/_mixins.scss",
      "src/core/public/chrome/ui/header/_index.scss",
      "src/core/public/chrome/ui/header/header_breadcrumbs.scss",
      "src/core/public/chrome/ui/header/home_loader.scss",
      "src/plugins/data/public/ui/filter_bar/_global_filter_item.scss",
      "src/plugins/home/public/application/components/_synopsis.scss",
      "src/plugins/vis_builder/public/application/components/searchable_dropdown.scss",
      "src/plugins/vis_builder/public/application/components/side_nav.scss",
      "packages/osd-ui-framework/src/components/button/button_group/_button_group.scss",
      "src/plugins/discover/public/application/components/data_grid/data_grid_table_cell_value.scss",
      "src/plugins/discover/public/application/view_components/canvas/discover_canvas.scss"
    ]
  }
}