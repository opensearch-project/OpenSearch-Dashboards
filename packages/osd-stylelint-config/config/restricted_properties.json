{
  "font-family": {
    "explanation": "All \"font-family\" styles should be inherited from OUI themes and components. Remove the rule.",
    "approved": [
      "src/plugins/maps_legacy/public/map/_leaflet_overrides.scss",
      "src/plugins/maps_legacy/public/map/_legend.scss",
      "src/plugins/opensearch_dashboards_legacy/public/font_awesome/font_awesome.scss",
      "src/plugins/opensearch_dashboards_react/public/markdown/_markdown.scss",
      "packages/osd-ui-framework/src/components/tool_bar/_tool_bar_search.scss",
      "packages/osd-ui-framework/src/global_styling/mixins/_global_mixins.scss",
      "src/plugins/data/public/ui/typeahead/_suggestion.scss",
      "src/plugins/vis_type_timeseries/public/application/components/_error.scss",
      "packages/osd-ui-framework/src/components/form/check_box/_check_box.scss",
      "src/plugins/discover/public/application/components/doc_viewer/doc_viewer.scss"
    ]
  }
}