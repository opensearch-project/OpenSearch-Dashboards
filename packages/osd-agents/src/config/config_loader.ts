/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/* eslint-disable no-console */

import { readFileSync, existsSync } from 'fs';
import { join } from 'path';
import { MCPServerConfig } from '../types/mcp_types';
import { Logger } from '../utils/logger';
import { BaseAGUIConfig } from '../ag_ui/base_ag_ui_adapter';

export class ConfigLoader {
  private static logger = new Logger();

  static async loadMCPConfig(): Promise<Record<string, MCPServerConfig>> {
    const configPath = join(__dirname, '../../configuration/mcp_config.json');

    if (existsSync(configPath)) {
      try {
        const configContent = readFileSync(configPath, 'utf-8');
        const configData = JSON.parse(configContent);
        const mcpConfigs = configData.mcpServers || configData;

        this.logger.info(`ðŸ“‹ Loaded MCP config from ${configPath}`, {
          configPath,
          serverCount: Object.keys(mcpConfigs).length,
        });
        console.log(`ðŸ“‹ Loaded MCP config from ${configPath}`);

        return mcpConfigs;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        this.logger.warn('Failed to load MCP config, using defaults', { error: errorMessage });
        console.warn('Failed to load MCP config, using defaults');
      }
    } else {
      this.logger.warn(`MCP config file not found at ${configPath}`);
    }

    return this.getDefaultMCPConfig();
  }

  static getDefaultMCPConfig(): Record<string, MCPServerConfig> {
    const defaultConfig = {
      filesystem: {
        type: 'local' as const,
        command: 'npx',
        args: ['-y', '@modelcontextprotocol/server-filesystem', '/tmp'],
      },
    };

    this.logger.info('Using default MCP configuration', defaultConfig);
    return defaultConfig;
  }

  static loadServerConfig(): BaseAGUIConfig {
    return {
      port: parseInt(process.env.AG_UI_PORT || '3001', 10),
      host: process.env.AG_UI_HOST || 'localhost',
      cors: {
        origins: process.env.AG_UI_CORS_ORIGINS?.split(',') || ['*'],
        credentials: process.env.AG_UI_CORS_CREDENTIALS === 'true',
      },
    };
  }
}
