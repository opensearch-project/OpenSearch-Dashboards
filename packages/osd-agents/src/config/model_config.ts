/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join } from 'path';
import { Logger } from '../utils/logger';

export interface ModelConfig {
  modelId: string;
}

export class ModelConfigManager {
  private static logger = new Logger();
  private static configPath = join(__dirname, '../../configuration/default-model.json');
  private static defaultModelId = 'us.anthropic.claude-sonnet-4-20250514-v1:0';

  static getDefaultModel(): ModelConfig {
    if (existsSync(this.configPath)) {
      try {
        const content = readFileSync(this.configPath, 'utf-8');
        const config = JSON.parse(content) as ModelConfig;
        this.logger.info('Loaded default model configuration', { modelId: config.modelId });
        return config;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        this.logger.warn('Failed to load default model config, using fallback', {
          error: errorMessage,
          fallback: this.defaultModelId,
        });
      }
    }

    return { modelId: this.defaultModelId };
  }

  static setDefaultModel(modelId: string): void {
    const config: ModelConfig = { modelId };

    try {
      writeFileSync(this.configPath, JSON.stringify(config, null, 2));
      this.logger.info('Updated default model configuration', { modelId });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      this.logger.error('Failed to save default model config', { error: errorMessage });
      throw new Error(`Failed to save model configuration: ${errorMessage}`);
    }
  }

  static resolveModelId(requestModelId?: string): string {
    if (requestModelId) {
      this.logger.debug('Using model from request', { modelId: requestModelId });
      return requestModelId;
    }

    const defaultConfig = this.getDefaultModel();
    this.logger.debug('Using default model', { modelId: defaultConfig.modelId });
    return defaultConfig.modelId;
  }
}
