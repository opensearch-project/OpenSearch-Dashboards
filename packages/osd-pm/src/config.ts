/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { resolve } from 'path';

interface Options {
  rootPath: string;
  skipOpenSearchDashboardsPlugins?: boolean;
  ossOnly?: boolean;
}

/**
 * Returns all the paths where plugins are located
 */
export function getProjectPaths({ rootPath, ossOnly, skipOpenSearchDashboardsPlugins }: Options) {
  const projectPaths = [rootPath, resolve(rootPath, 'packages/*')];

  // This is needed in order to install the dependencies for the declared
  // plugin functional used in the selenium functional tests.
  // As we are now using the webpack dll for the client vendors dependencies
  // when we run the plugin functional tests against the distributable
  // dependencies used by such plugins like @eui, react and react-dom can't
  // be loaded from the dll as the context is different from the one declared
  // into the webpack dll reference plugin.
  // In anyway, have a plugin declaring their own dependencies is the
  // correct and the expect behavior.
  projectPaths.push(resolve(rootPath, 'test/plugin_functional/plugins/*'));
  projectPaths.push(resolve(rootPath, 'test/interpreter_functional/plugins/*'));
  projectPaths.push(resolve(rootPath, 'examples/*'));

  if (!skipOpenSearchDashboardsPlugins) {
    projectPaths.push(resolve(rootPath, '../opensearch-dashboards-extra/*'));
    projectPaths.push(resolve(rootPath, '../opensearch-dashboards-extra/*/packages/*'));
    projectPaths.push(resolve(rootPath, '../opensearch-dashboards-extra/*/plugins/*'));
    projectPaths.push(resolve(rootPath, 'plugins/*'));
    projectPaths.push(resolve(rootPath, 'plugins/*/packages/*'));
    projectPaths.push(resolve(rootPath, 'plugins/*/plugins/*'));
  }

  return projectPaths;
}
