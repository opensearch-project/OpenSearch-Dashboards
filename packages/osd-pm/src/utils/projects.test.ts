/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { mkdir, symlink } from 'fs';
import { join, resolve } from 'path';
import del from 'del';
import { promisify } from 'util';

import { getProjectPaths } from '../config';
import { Project } from './project';
import {
  buildProjectGraph,
  getProjects,
  includeTransitiveProjects,
  ProjectGraph,
  ProjectMap,
  topologicallyBatchProjects,
} from './projects';
import { PROCESS_WORKING_DIR } from '@osd/cross-platform';

const rootPath = resolve(__dirname, '__fixtures__/opensearch-dashboards');
const rootPlugins = join(rootPath, 'plugins');

describe('#getProjects', () => {
  beforeAll(async () => {
    // Make sure we start clean
    await del(rootPlugins, { cwd: PROCESS_WORKING_DIR });

    await promisify(mkdir)(rootPlugins);

    await promisify(symlink)(
      join(__dirname, '__fixtures__/symlinked-plugins/corge'),
      join(rootPlugins, 'corge'),
      'junction' // This parameter would only be used on Windows
    );
  });

  afterAll(async () => await del(rootPlugins, { cwd: PROCESS_WORKING_DIR }));

  test('find all packages in the packages directory', async () => {
    const projects = await getProjects(rootPath, ['packages/*']);

    const expectedProjects = ['bar', 'foo'];

    expect(projects.size).toBe(2);
    expect([...projects.keys()]).toEqual(expect.arrayContaining(expectedProjects));
  });

  test('can specify root as a separate project', async () => {
    const projects = await getProjects(rootPath, ['.']);

    expect(projects.size).toBe(1);
    expect([...projects.keys()]).toEqual(['opensearch-dashboards']);
  });

  test('handles packages outside root', async () => {
    const projects = await getProjects(rootPath, ['../plugins/*']);

    const expectedProjects = ['baz', 'quux', 'zorge'];

    expect(projects.size).toBe(3);
    expect([...projects.keys()]).toEqual(expect.arrayContaining(expectedProjects));
  });

  test('throws if multiple projects has the same name', async () => {
    await expect(
      getProjects(rootPath, ['../plugins/*', '../other-plugins/*'])
    ).rejects.toHaveProperty('message', 'There are multiple projects with the same name [baz]');
  });

  test('includes additional projects in package.json', async () => {
    const projectPaths = getProjectPaths({ rootPath });
    const projects = await getProjects(rootPath, projectPaths);

    const expectedProjects = [
      'opensearch-dashboards',
      'bar',
      'foo',
      'with-additional-projects',
      'quux',
      'baz',
      'bar',
    ];

    expect([...projects.keys()]).toEqual(expect.arrayContaining(expectedProjects));
    expect(projects.size).toBe(expectedProjects.length);
  });

  describe('with exclude/include filters', () => {
    let projectPaths: string[];
    beforeEach(() => {
      projectPaths = getProjectPaths({ rootPath });
    });

    test('excludes projects specified in `exclude` filter', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        exclude: ['foo', 'bar', 'baz'],
      });

      expect([...projects.keys()].sort()).toEqual([
        'corge',
        'opensearch-dashboards',
        'quux',
        'with-additional-projects',
      ]);
    });

    test('ignores unknown projects specified in `exclude` filter', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        exclude: ['unknown-foo', 'bar', 'unknown-baz'],
      });

      expect([...projects.keys()].sort()).toEqual([
        'baz',
        'corge',
        'foo',
        'opensearch-dashboards',
        'quux',
        'with-additional-projects',
      ]);
    });

    test('includes only projects specified in `include` filter', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        include: ['foo', 'bar'],
      });

      expect([...projects.keys()].sort()).toEqual(['bar', 'foo']);
    });

    test('ignores unknown projects specified in `include` filter', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        include: ['unknown-foo', 'bar', 'unknown-baz'],
      });

      expect([...projects.keys()].sort()).toEqual(['bar']);
    });

    test('respects both `include` and `exclude` filters if specified at the same time', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        exclude: ['bar'],
        include: ['foo', 'bar', 'baz'],
      });

      expect([...projects.keys()].sort()).toEqual(['baz', 'foo']);
    });

    test('does not return any project if wrong `include` filter is specified', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        include: ['unknown-foo', 'unknown-bar'],
      });

      expect(projects.size).toBe(0);
    });

    test('does not return any project if `exclude` filter is specified for all projects', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        exclude: [
          'opensearch-dashboards',
          'bar',
          'corge',
          'foo',
          'with-additional-projects',
          'quux',
          'baz',
        ],
      });

      expect(projects.size).toBe(0);
    });

    test('does not return any project if `exclude` and `include` filters are mutually exclusive', async () => {
      const projects = await getProjects(rootPath, projectPaths, {
        exclude: ['foo', 'bar'],
        include: ['foo', 'bar'],
      });

      expect(projects.size).toBe(0);
    });
  });
});

describe('#buildProjectGraph', () => {
  test('builds full project graph', async () => {
    const allProjects = await getProjects(rootPath, ['.', 'packages/*', '../plugins/*']);
    const graph = buildProjectGraph(allProjects);

    const expected: { [k: string]: string[] } = {};
    for (const [projectName, projects] of graph.entries()) {
      expected[projectName] = projects.map((project: Project) => project.name);
    }

    expect(expected).toMatchSnapshot();
  });
});

describe('#topologicallyBatchProjects', () => {
  let projects: ProjectMap;
  let graph: ProjectGraph;
  beforeEach(async () => {
    projects = await getProjects(rootPath, ['.', 'packages/*', '../plugins/*']);
    graph = buildProjectGraph(projects);
  });

  test('batches projects topologically based on their project dependencies', async () => {
    const batches = topologicallyBatchProjects(projects, graph);

    const expectedBatches = batches.map((batch) => batch.map((project) => project.name));

    expect(expectedBatches).toMatchSnapshot();
  });

  test('batches projects topologically even if graph contains projects not presented in the project map', async () => {
    // Make sure that the project we remove really existed in the projects map.
    expect(projects.delete('foo')).toBe(true);

    const batches = topologicallyBatchProjects(projects, graph);

    const expectedBatches = batches.map((batch) => batch.map((project) => project.name));

    expect(expectedBatches).toMatchSnapshot();
  });

  describe('batchByWorkspace = true', () => {
    test('batches projects topologically based on their project dependencies and workspaces', async () => {
      const batches = topologicallyBatchProjects(projects, graph, { batchByWorkspace: true });

      const expectedBatches = batches.map((batch) => batch.map((project) => project.name));

      expect(expectedBatches).toEqual([
        ['opensearch-dashboards'],
        ['bar', 'foo'],
        ['baz', 'zorge'],
        ['quux'],
      ]);
    });
  });
});

describe('#includeTransitiveProjects', () => {
  test('includes transitive dependencies for OpenSearch Dashboards package', async () => {
    const projects = await getProjects(rootPath, ['.', 'packages/*']);

    const opensearchDashboards = projects.get('opensearch-dashboards')!;
    const withTransitive = includeTransitiveProjects([opensearchDashboards], projects);

    expect([...withTransitive.keys()]).toEqual(['opensearch-dashboards', 'foo']);
  });

  test('handles multiple projects with same transitive dep', async () => {
    const projects = await getProjects(rootPath, ['.', 'packages/*']);

    const opensearchDashboards = projects.get('opensearch-dashboards')!;
    const bar = projects.get('bar')!;
    const withTransitive = includeTransitiveProjects([opensearchDashboards, bar], projects);

    expect([...withTransitive.keys()]).toEqual(['opensearch-dashboards', 'bar', 'foo']);
  });

  test('handles projects with no deps', async () => {
    const projects = await getProjects(rootPath, ['.', 'packages/*']);

    const foo = projects.get('foo')!;
    const withTransitive = includeTransitiveProjects([foo], projects);

    expect([...withTransitive.keys()]).toEqual(['foo']);
  });

  test('includes dependencies of dependencies', async () => {
    const projects = await getProjects(rootPath, ['.', 'packages/*', '../plugins/*']);

    const quux = projects.get('quux')!;
    const withTransitive = includeTransitiveProjects([quux], projects);

    expect([...withTransitive.keys()]).toEqual(['quux', 'bar', 'baz', 'foo']);
  });
});
