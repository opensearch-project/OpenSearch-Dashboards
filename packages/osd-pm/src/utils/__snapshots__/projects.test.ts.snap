// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildProjectGraph builds full project graph 1`] = `
Object {
  "bar": Array [
    "foo",
  ],
  "baz": Array [],
  "foo": Array [],
  "opensearch-dashboards": Array [
    "foo",
  ],
  "quux": Array [
    "bar",
    "baz",
  ],
  "zorge": Array [
    "foo",
  ],
}
`;

exports[`#topologicallyBatchProjects batches projects topologically based on their project dependencies 1`] = `
Array [
  Array [
    "foo",
    "baz",
  ],
  Array [
    "opensearch-dashboards",
    "bar",
    "zorge",
  ],
  Array [
    "quux",
  ],
]
`;

exports[`#topologicallyBatchProjects batches projects topologically even if graph contains projects not presented in the project map 1`] = `
Array [
  Array [
    "opensearch-dashboards",
    "bar",
    "baz",
    "zorge",
  ],
  Array [
    "quux",
  ],
]
`;
